androidtest/memtag/memtag_test.cc

```diff
@@ -0,0 +1,204 @@
+// needed to uncondionally enable assertions
+#undef NDEBUG
+#include <assert.h>
+#include <malloc.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <sys/utsname.h>
+#include <unistd.h>
+
+#include <map>
+#include <string>
+#include <unordered_map>
+
+using namespace std;
+
+using u8 = uint8_t;
+using uptr = uintptr_t;
+using u64 = uint64_t;
+
+const size_t DEFAULT_ALLOC_SIZE = 8;
+const size_t CANARY_SIZE = 8;
+
+void do_context_switch() {
+    utsname s;
+    uname(&s);
+}
+
+u8 get_pointer_tag(void *ptr) {
+    return (((uptr) ptr) >> 56) & 0xf;
+}
+
+void *untag_pointer(void *ptr) {
+    const uintptr_t mask = UINTPTR_MAX >> 8;
+    return (void *) ((uintptr_t) ptr & mask);
+}
+
+void tag_distinctness() {
+    if (rand() & 1) {
+        // make allocations in all of used size classes and free half of them
+
+        const int max = 21000;
+        void *ptrs[max];
+
+        for (int i = 0; i < max; ++i) {
+            ptrs[i] = malloc(max);
+        }
+
+        for (int i = 1; i < max; i += 2) {
+            free(ptrs[i]);
+        }
+    }
+
+    const size_t cnt = 3000;
+    const size_t iter_cnt = 5;
+    const size_t alloc_cnt = cnt * iter_cnt;
+
+    const int sizes[] = { 16, 160, 10240, 20480 };
+
+    for (size_t size_idx = 0; size_idx < sizeof(sizes) / sizeof(int); ++size_idx) {
+        const size_t full_alloc_size = sizes[size_idx];
+        const size_t alloc_size = full_alloc_size - CANARY_SIZE;
+
+        unordered_map<uptr, u8> map;
+        map.reserve(alloc_cnt);
+
+        for (size_t iter = 0; iter < iter_cnt; ++iter) {
+            uptr allocations[cnt];
+
+            for (size_t i = 0; i < cnt; ++i) {
+                u8 *p = (u8 *) malloc(alloc_size);
```

Comment:
```suggestion
                u8 *p = (u8 *) malloc(alloc_size);
                assert(p);
```

Reply:
Addressed a s part of https://github.com/GrapheneOS/hardened_malloc/pull/225/commits/072fd5b9779ec8900d619799953ee0ea5a58da47